das/state.go

```diff
@@ -201,28 +202,31 @@ func (s *coordinatorState) unsafeStats() SamplingStats {
 		Failed:           failed,
 		Workers:          workers,
 		Concurrency:      len(workers),
-		CatchUpDone:      s.catchUpDone,
-		IsRunning:        len(workers) > 0 || s.catchUpDone,
+		CatchUpDone:      s.catchUpDone.Load(),
+		IsRunning:        len(workers) > 0 || s.catchUpDone.Load(),
 	}
 }
 
 func (s *coordinatorState) checkDone() {
 	if len(s.inProgress) == 0 && len(s.priority) == 0 && s.next > s.networkHead {
-		if !s.catchUpDone {
+		if s.catchUpDone.CompareAndSwap(false, true) {
 			close(s.catchUpDoneCh)
-			s.catchUpDone = true
 		}
 		return
 	}
 
-	if s.catchUpDone {
+	if s.catchUpDone.Load() {
+		// overwrite channel before storing done flag
 		s.catchUpDoneCh = make(chan struct{})
-		s.catchUpDone = false
+		s.catchUpDone.Store(false)
 	}
 }
 
 // waitCatchUp waits for sampling process to indicate catchup is done
 func (s *coordinatorState) waitCatchUp(ctx context.Context) error {
+	if s.catchUpDone.Load() {
+		return nil
+	}
 	select {
 	case <-s.catchUpDoneCh:
```

Comment:
There is still can be a race condition for `catchUpDoneCh`. Because we rewrite the actual chan.

Reply:
It is not possible, because overwrite is by atomic.Bool var. read could be done only if catchUpDone == false. Write could be done only if catchUpDone == true, when read is short-circuit. So to overwrite it, we make sure it is catchUpDone == true, then do overwrite and only then allow reads, by setting catchUpDone.Store(false).

Reply:
On the other hand, closing the channel is concurrently same operation, so we can use CAS instead of 2 operations.

Reply:
I see, missed the short-circuit part.

Reply:
Should we store the channel in the atomic value instead?

Reply:
Then we can avoid a bool fully.

Reply:
Just a suggestion and this full up to you